Update angular-bootstrap DatePicker with UI Framework classes.

cjcenizal opened the pull request:
## Changes

* Add DatePicker component to UI Framework.
* Update angular-bootstrap date-picker to use these classes.
* Improve DatePicker dark-theme.
* Remove dependency upon Bootstrap buttons.

## Component

![image](https://cloud.githubusercontent.com/assets/1238659/25307645/593d7a58-2759-11e7-899d-e29786e9f948.png)

## Before

![image](https://cloud.githubusercontent.com/assets/1238659/25305547/16fba8d0-2732-11e7-9d07-f12699adcfda.png)

![image](https://cloud.githubusercontent.com/assets/1238659/25305549/1985edb8-2732-11e7-99d6-2a19b58f6afe.png)

![image](https://cloud.githubusercontent.com/assets/1238659/25307613/d9bb0106-2758-11e7-8f8f-42c9afda7eea.png)

![image](https://cloud.githubusercontent.com/assets/1238659/25307614/db496fe4-2758-11e7-9921-985b58eaa6f8.png)

![image](https://cloud.githubusercontent.com/assets/1238659/25307615/dd60c23c-2758-11e7-926a-9bca70aec947.png)

![image](https://cloud.githubusercontent.com/assets/1238659/25307616/df850e38-2758-11e7-8678-4604c961bf98.png)

## After

![image](https://cloud.githubusercontent.com/assets/1238659/25305551/1d0fb2d4-2732-11e7-93e5-d4bab2cea8c4.png)

![image](https://cloud.githubusercontent.com/assets/1238659/25305552/2036176e-2732-11e7-8e6c-548bbd324ece.png)

![image](https://cloud.githubusercontent.com/assets/1238659/25307597/9b4f7c08-2758-11e7-8ade-c5a74afa2940.png)

![image](https://cloud.githubusercontent.com/assets/1238659/25307598/9df0bd46-2758-11e7-95a2-133d7e8e1288.png)

![image](https://cloud.githubusercontent.com/assets/1238659/25307601/a1afca76-2758-11e7-941b-6dd493a65ece.png)

![image](https://cloud.githubusercontent.com/assets/1238659/25307602/a41773ae-2758-11e7-82eb-7cf2d5b66979.png)

kobelb commented on src/ui/public/angular-bootstrap/datepicker/month.html:
```diff
@@ -1,15 +1,72 @@
+            class="kuiDatePickerNavigationButton"
+            ng-click="toggleMode()"
+            tabindex="-1"
+            style="width:100%;"
```
Are we comfortable with this inline-style even after the conversion to the ui-framework?

cjcenizal replied:
Great catch, I'll remove those.

kobelb commented on src/ui/public/angular-bootstrap/datepicker/day.html:
```diff
@@ -1,20 +1,82 @@
+      <td
+        ng-repeat="dt in row track by dt.date"
+        class="kuiDatePickerRowCell"
+        ng-class="{'kuiDatePickerRowCell-isPrecluded': dt.disabled && !dt.secondary}"
```
Question regarding the `isPrecluded` naming, what made you choose it over `isDisabled` which seems to be common throughout our codebase.

cjcenizal replied:
The angular-bootstrap calendar is weird, in that you can't select a day on the "From" calendar if it falls after the selected day in the "To" calendar (and vice versa, you can't select a "To" day if it is before the "From" day). So in this sense, those unavailable days are "precluded" by the selection on the other calendar. I thought it was important to distinguish this from a simple "disabled" state since you can actually disable days for other reasons (e.g. if a day falls on the previous or following month). So that's why I chose "precluded". What do you think? Does this name communicate the intent, or can you think of a name that makes more sense? Open to suggestions. :)

kobelb replied:
Ah, that makes sense why you'd want to use a different name for that situation. Honestly, I didn't pick up the intent of the word and resorted to using the dictionary to help me out, and still came up empty handed. How do you feel about "isConstrained" or "isBlocked"? Don't let my poor vocabulary purely impact your naming decision though :)

cjcenizal replied:
isBlocked is good! Thanks for the suggestion. I'll update the code.

snide commented on ui_framework/components/local_nav/_index.scss:
```diff
@@ -17,6 +17,12 @@ $localTabTextColor-isSelected: $localNavButtonTextColor-isHover;
 $localSearchButtonBackgroundColor: #9c9c9c;
 $localSearchButtonTextColor: #ffffff;
+$localCalendarButtonHoverBackgroundColor: $localNavButtonBackgroundColor-isHover;
+$localCalendarCellPrecludedBackgroundColor: rgba(black, 0.05);
```
I think you can loose this from a design standpoint. The lighter text is enough. You might want to put a `cursor: not-allowed` on the elements precluded.

snide commented on ui_framework/components/local_nav/_index.scss:
```diff
@@ -17,6 +17,12 @@ $localTabTextColor-isSelected: $localNavButtonTextColor-isHover;
+$localCalendarCellPrecludedBackgroundColor: rgba(black, 0.05);
+$localCalendarCellSelectedBackgroundColor: #777777;
+$localCalendarCellSelectedTextColor: #ffffff;
+$localCalendarCellHoverBackgroundColor: $globalInfoColor;
```
![image](https://d3vv6lp55qjaqc.cloudfront.net/items/3S3p3T0P0M0Z1a3d2R0H/Screen%20Recording%202017-04-24%20at%2011.35%20AM.gif?X-CloudApp-Visitor-Id=59773&v=bf91f058)

The hover state of the cal and the nav within the cal are different. I'd pick one and stick with it.

snide replied:
Hover state should almost always trump a selected state.

![](https://d3vv6lp55qjaqc.cloudfront.net/items/3B42081c1D3F3d0D231e/Screen%20Recording%202017-04-24%20at%2011.40%20AM.gif?X-CloudApp-Visitor-Id=59773&v=62c804c3)

cjcenizal replied:
What's the rationale behind this rule? A selected day can't be selected again, so it seems like we would want to communicate that visually?

snide replied:
If that's the case, then use a cursor:not-allowed. I assumed it was because of the pointer.

cjcenizal replied:
Oh actually I was wrong. You _can_ select it again, which will reset the time. I'll make the change you suggested, thanks.

snide commented on ui_framework/components/local_nav/_local_calendar.scss:
```diff
@@ -0,0 +1,145 @@
+
+  &.kuiDatePickerRowCellContent-isCurrent {
+    color: $localNavTextColor; /* 1 */
+    border: 1px dotted $globalSubduedTextColor;
```
Don't like the dotted here. It's another boxy element (4 of them). I'd go with the color of the background hover state (blue) for the text, and leave the rest of the design elements off. Similar to the original design.

![image](https://cloud.githubusercontent.com/assets/324519/25353248/7255db58-28e3-11e7-9a98-c20f04bf0c1a.png)

cjcenizal replied:
I'll give it a shot, I'm just concerned about the contrast.

snide commented on src/ui/public/angular-bootstrap/datepicker/year.html:
```diff
@@ -1,15 +1,72 @@
-<table role="grid" aria-labelledby="{{uniqueId}}-title" aria-activedescendant="{{activeDateId}}">
+<table
```
On the year selector I'd try to get your alignment correct (properly gridded). This is likely happening because your spacing for the day/date doesn't have the padding that you're putting on the "buttons". Because the year selector doesn't have the day name thead element it gets misaligned.

![image](https://cloud.githubusercontent.com/assets/324519/25353468/2d8964a8-28e4-11e7-9386-17ac9b29e3cb.png)

uboness replied:
the selected range is not the most obvious here.. is there a reason not to highlight the range with the same background as the bounding dates?

cjcenizal replied:
I agree, but I don't think angular-bootstrap's datepicker directive surfaces this information. We have to update the logic of this directive to support highlighting the current range. I'd also like to get rid of the silly way the datepicker highlights the days you _can't_ support. Dave's working on a new datepicker design, so I think we can address these problems when we retrofit that design.

cjcenizal commented:
@snide Addressed your feedback and ready for another review.

Alignment:

![image](https://cloud.githubusercontent.com/assets/1238659/25355936/49f17974-28ed-11e7-85d9-cc037b66bf9c.png)

Visual states:

![datepicker](https://cloud.githubusercontent.com/assets/1238659/25355960/5b7bd4fa-28ed-11e7-894c-bf496ed35607.gif)

![datepicker_dark](https://cloud.githubusercontent.com/assets/1238659/25355962/5e42face-28ed-11e7-8fd0-ee37b09168c2.gif)
